FIX Don't use int width for mysql >= 8.0.17 #9453

gurucomkz opened the pull request:
Fixes issue #9453:

Added function `shouldUseIntegerWidth()` to `MySQLSchemaManager` to check for MySQL version and used it in functions like `int()`, `bigint()`, `tinyint()` to determine whether the width attribute should be used for this mysql version.

Achieves the result described in the task.

michalkleiner commented:
Without knowing what is or isn't available, I think the version check needs to be stronger. How does e.g. MariDB 10.x report the version? Would it become unintentionally affected by this tweak too?

gurucomkz commented:
@michalkleiner Good point. Let me check that.

gurucomkz commented:
@michalkleiner Okay, MariaDB doesn't discard the width parameter at all. I've added a condition to detect MariaDB and act accordingly. I have no idea how strictly Percona follows MySQL changes, so maybe it may worth adding a config flag that will override the auto-discovery?

sminnee commented:
You could have a config setting for this that can be true/false/null.

Null is the default, and will auto detect based on version number for MySQL and MariaDB.

But can be set to true or false to force the behaviour if the auto detector fails.

gurucomkz commented:
@sminnee Thank you for your feedback. I've added the config flag + a note in the docs.

What else should I do to make this pull request considered?

dhensby reviewed (approved):
This looks good to me, any other feedback from @silverstripe/core-team ?

Cheddam commented:
The code looks good to me, but the changelog notes will need to be shifted, as 4.7.0 has well and truly shipped.

@gurucomkz If you mark the newly added method as `private`, I think it's safe to land this in 4.8.0 (which is currently in beta) by moving the changelog notes to 4.8.0.md and changing the branch target to `4.8`. Sorry for the runaround on this PR - excited to see it shipped, though!

maxime-rainville commented:
4.8-rc1 is shipping on Monday. We won't want to include this in the RC. Target the change log entry to 4.9.

maxime-rainville commented:
Ended up going down a rabbit hole to test this in my local environment. Here's how you can use [MySQL-server 8 with Docker](https://forum.silverstripe.org/t/running-a-development-site-with-mysql-8-in-docker/4100) to test this PR.

gurucomkz commented:
Sorry guys, I’m stuck in a hospital for the week and can’t participate. If anyone has time, please, adjust the PR. Thank you for your attention!

maxime-rainville commented on src/ORM/Connect/MySQLSchemaManager.php:
```diff
@@ -241,6 +241,23 @@ public function renameField($tableName, $oldName, $newName)
 
     protected static $_cache_collation_info = [];
 
+    public function shouldUseIntegerWidth()
```
Ideally we would have some unit test covering. It might be possible to do it with some semi-elaborate mock.

It's probably good enough as it is.

gurucomkz replied:
Done

maxime-rainville reviewed (approved):
Tested this locally with my Docker set up. I tried creating a field using various `int` types and swapping between them to make sure things behaved as expected.

I've copied the change log snippet to the regular doc and move it to the 4.9 changelog. I've rebase the PR to be in sink with the HEAD of `4`.

I'll hold off merging to give a chance to other people to chip in. But we got a couple core commitors approving of this ... so if anyone has beef with this PR, say it soonish.

gurucomkz commented:
I'm sorry for taking so long to get back to your comments @maxime-rainville .

I've changed the function's visibility and also added a test to demonstrate the functionality on a simulated `Database::getVersion()` calls.

michalkleiner commented on docs/en/00_Getting_Started/00_Server_Requirements.md:
```diff
@@ -45,6 +45,17 @@ Existing projects that upgrade to Recipe 4.7.0 will unintentionally adopt this c
+ ### MySQL/MariaDB Int width in schema
+
+MySQL 8.0.17 stopped reporting the width attribute for integers while MariaDB did not change its behaviour.
+This results in constant rebuilding the schema when MySQLSchemaManager expects a field to look like i.e.
```
```suggestion
This results in constant rebuilding of the schema when MySQLSchemaManager expects a field to look like e.g.
```

michalkleiner commented on docs/en/04_Changelogs/4.9.0.md:
```diff
@@ -0,0 +1,15 @@
+
+### MySQL/MariaDB Int width in schema
+MySQL 8.0.17 stopped reporting the width attribute for integers while MariaDB did not change its behaviour. 
+This results in constant rebuilding the schema when MySQLSchemaManager expects a field to look like i.e. 
```
```suggestion
This results in constant rebuilding of the schema when MySQLSchemaManager expects a field to look like e.g. 
```

michalkleiner commented on docs/en/04_Changelogs/4.9.0.md:
```diff
@@ -0,0 +1,15 @@
+# 4.9.0 (Unreleased)
```
4.9.0 is now out, so we probably need to target `4.9` branch to make this into `4.9.1` and update the changelog accordingly, is that right @maxime-rainville?

michalkleiner replied:
I think this can be treated as a bugfix — it fixes broken/unexpected behaviour, so I think it doesn't need to be treated as a new feature.

gurucomkz commented:
Rebased on top of the latest code and resolved the conflicts. I hope it helps with including this PR in some of the next releases.

maxime-rainville commented on docs/en/04_Changelogs/4.9.0.md:
````diff
@@ -249,6 +249,16 @@ New Silverstripe CMS project created from `silverstripe/installer` 4.9.0 or grea
 
 It's highly recommended that existing Silverstripe CMS installation using `Swift_MailTransport` upgrade to `Swift_SendmailTransport` or another available transport, such as `Swift_SmtpTransport`. Details on how to use these classes are available in the [email section](https://docs.silverstripe.org/en/4/developer_guides/email/) of the developer docs.
 
+### MySQL/MariaDB Int width in schema
+MySQL 8.0.17 stopped reporting the width attribute for integers while MariaDB did not change its behaviour. 
+This results in constant rebuilding of the schema when MySQLSchemaManager expects a field to look like e.g. 
+`INT(8)` and MySQL server reports it simply as `INT`. MySQLSchemaManager has been updated to detect the MySQL
+server implementation and act accordingly. In cases when auto-detection fails, you can force the desired behaviour like this:
+```yml
+SilverStripe\ORM\Connect\MySQLSchemaManager:
+    schema_use_int_width: true # or false when INT widths should be ignored
+```
+
````
```suggestion
```

maxime-rainville reviewed (changes requested):
Get this bit out of the change log and we're good to merge.

maxime-rainville reviewed (approved):
LGTM
